Practice Exercise

In [1]:
#import of libraries
import pandas as pd
import numpy as np
import os

In [2]:
df = pd.read_csv(r'/Users/tiffanib./Desktop/Career Foundry Data Analytics/Achievement 4 - Python/28-09-2025 Instacart Basket Analysis/02 Data/Original Data/orders.csv',index_col = False)

In [3]:
path = r'/Users/tiffanib./Desktop/Career Foundry Data Analytics/Achievement 4 - Python/28-09-2025 Instacart Basket Analysis'

In [4]:
path

'/Users/tiffanib./Desktop/Career Foundry Data Analytics/Achievement 4 - Python/28-09-2025 Instacart Basket Analysis'

In [5]:
#import df_prods
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [6]:
#import df_ords
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [7]:
#Create a dataframe
df_test = pd.DataFrame()

In [9]:
df_test['mix'] = ['a','b',1,True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [22]:
#check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)


mix


In [23]:
df_test['mix']=df_test['mix'].astype('str')

In [24]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [25]:
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [26]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [27]:
df_prods.shape

(49693, 5)

In [28]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [29]:
df_prods_clean.shape

(49677, 5)

In [30]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [31]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [32]:
df_prods_clean.shape

(49677, 5)

In [33]:
#creating database tht doesnt have duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [35]:
df_prods_clean_no_dups.shape

(49672, 5)

Task 4.5 of Data Consistency Checks (1-10)

Number 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

Number 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [36]:
df_ords.describe()

The history saving thread hit an unexpected error (OperationalError('attempt to write a readonly database')).History will not be written to the database.


Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  orders_hour_of_day  days_since_prior_order  
count        3.421083e+06        3.421083e+06            3.214874e+06  
mean         2.776219e+00        1.345202e+01            1.111484e+01  
std          2.046829e+00        4.226088e+00            9.206737e+00  
min          0.000000e+00        0.000000e+00            0.000000e+00  
25%          1.000000e+00        1.000000e+01            4.000000e+00  
50%          3.000000e+00        1.300000e+01            7.000000e+00  
75%          5.000000e+00        1.600000e+01            1.500000e+01  
max          6.000000e+00        2.300000e+01            3.000000e+01

For orders_day_of_the_week - The minimum an maximum range is 0 ad 6. 0 for Monday and 6 for Saturday. The max value of 100 is for the order_number. For Order_hour_of_day the count appears to be incorrect. 

Number 3. Check for mixed-type data in your df_ords dataframe.

In [37]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)


/var/folders/5p/rxp_8kqn1tgbtc0fc968b0mh0000gn/T/ipykernel_81861/1819069806.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)


In [38]:
#the code is showing incorrect. Making changes to accomdate.
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].map(type)).any(axis=1)
    if len(df_test[weird]) > 0:
        print(col)


In [39]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
orders_hour_of_day          int64
days_since_prior_order    float64
dtype: object

Number 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no columns with mixed data types.

Number 5. Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [40]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
orders_hour_of_day             0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the days_since_prior_order column. This could be due to first time customers who are placing their orders for the first time. There is no prior order to compare to.

Number 6. Address the missing values using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [41]:
#checking to see if these missing values are only within the days since prior order column
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [42]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         orders_hour_of_day  days_since_prior_order  
0                         8                     NaN  
11                       11                     NaN  
26                       14                     NaN  
39                       11                     NaN  
45                       12                     NaN  
...                     ...                     ...  
3420930                  12                     NaN  
3420934                  18                     NaN  
3421002                  19                     NaN  
3421019                  15                     NaN  
3421069                  11                     NaN  

[206209 rows x 7 columns]

In [43]:
#checking to see if this is their first order
df_ords_first_order = df_ords_nan[df_ords_nan['order_number']==1]

In [44]:
df_ords_first_order

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         orders_hour_of_day  days_since_prior_order  
0                         8                     NaN  
11                       11                     NaN  
26                       14                     NaN  
39                       11                     NaN  
45                       12                     NaN  
...                     ...                     ...  
3420930                  12                     NaN  
3420934                  18                     NaN  
3421002                  19                     NaN  
3421019                  15                     NaN  
3421069                  11                     NaN  

[206209 rows x 7 columns]

In [45]:
# this shows that it is the customers first order. The rows and columns match

Number 7. Run a check for duplicate values in your df_ords data.
In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [47]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [48]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, orders_hour_of_day, days_since_prior_order]
Index: []

There are no duplicates values.

Number 8. Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

There are no duplicates found.

Number 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [49]:
df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))

In [51]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'product_cleaned.csv'))